### PR TITLE
add test for val split

### LINKED ISSUE #203: val_split returns object of type 'NoneType' has no len()
## 🐛 Bug

When using the branch [`base_viz_2`](https://github.com/PyTorchLightning/lightning-flash/tree/base_viz_2) and I pass `val_split` to `ImageClassificationData.from_filepaths` I get the follow error:
```python
# 2. Load the data

datamodule = ImageClassificationData.from_filepaths(
    train_filepaths=train_images,
    train_labels=train_labels,
    batch_size=32,
    num_workers=mproc.cpu_count(),
    train_transform=TRAIN_TRANSFORM,
    val_transform=VALID_TRANSFORM,
    val_split=0.3  # crashes with base_viz_2
)
```
```
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-30-01c3b871d89f> in <module>
      8     train_transform=TRAIN_TRANSFORM,
      9     val_transform=VALID_TRANSFORM,
---> 10     val_split=0.3  # crashes with base_viz_2
     11 )

/opt/conda/lib/python3.7/site-packages/flash/vision/classification/data.py in from_filepaths(cls, train_filepaths, train_labels, val_filepaths, val_labels, test_filepaths, test_labels, predict_filepaths, train_transform, val_transform, test_transform, predict_transform, batch_size, num_workers, seed, preprocess_cls, **kwargs)
    529             preprocess=preprocess,
    530             seed=seed,
--> 531             **kwargs
    532         )

/opt/conda/lib/python3.7/site-packages/flash/data/data_module.py in from_load_data_inputs(cls, train_load_data_input, val_load_data_input, test_load_data_input, predict_load_data_input, preprocess, postprocess, **kwargs)
    375             data_pipeline = DataPipeline(
    376                 preprocess or cls(**kwargs).preprocess,
--> 377                 postprocess or cls(**kwargs).postprocess,
    378             )
    379         else:

/opt/conda/lib/python3.7/site-packages/pytorch_lightning/core/datamodule.py in __call__(cls, *args, **kwargs)
     48 
     49         # Get instance of LightningDataModule by mocking its __init__ via __call__
---> 50         obj = type.__call__(cls, *args, **kwargs)
     51 
     52         return obj

/opt/conda/lib/python3.7/site-packages/flash/vision/classification/data.py in __init__(self, train_dataset, val_dataset, test_dataset, predict_dataset, batch_size, num_workers, seed, train_split, val_split, test_split, **kwargs)
    214         if train_dataset is not None and train_split is not None or val_split is not None or test_split is not None:
    215             train_dataset, val_dataset, test_dataset = self.train_val_test_split(
--> 216                 train_dataset, train_split, val_split, test_split, seed
    217             )
    218 

/opt/conda/lib/python3.7/site-packages/flash/data/data_module.py in train_val_test_split(dataset, train_split, val_split, test_split, seed)
    294 
    295         """
--> 296         n = len(dataset)
    297 
    298         if test_split is None:

TypeError: object of type 'NoneType' has no len()
```

#### Code sample

https://colab.research.google.com/drive/1OHTFzWEock7lQ24SkD4oGDhdDkcU_mlf?usp=sharing

### Environment

In addition to that, in a Kaggle notebook environment the performance dropped in terms of 1-2min when I create an instance of `ImageClassificationData`.

### LINKED ISSUE #235: `ImageClassificationPreprocess.image_size` not being saved with the checkpoint
## 🐛 Bug

If you create an ImageClassificationPreprocess and set `.image_size` to something custom, when the checkpoint is saved and reloaded the image size will have reverted to the default.

### Additional context

Pickle seems to have trouble with attributes defined on the class rather than in the `__init__`. This may have an impact for more than just `image_size` as we do this a lot.

### PR DESCRIPTION
**Pull request by edgarriba**
## What does this PR do?

Fixes #203 
Fixes #235 

## Before submitting
- [ ] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [ ] Did you read the [contributor guideline](https://github.com/PyTorchLightning/lightning-flash/tree/master/.github/CONTRIBUTING.md), Pull Request section?
- [ ] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [ ] Did you make sure to update the documentation with your changes?
- [ ] Did you write any new necessary tests? [not needed for typos/docs]
- [ ] Did you verify new and existing tests pass locally with your changes?
- [ ] If you made a notable change (that affects users), did you update the [CHANGELOG](https://github.com/PyTorchLightning/lightning-flash/blob/master/CHANGELOG.md)?

## PR review
 - [ ] Is this pull request ready for review? (if not, please submit in draft mode)

Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: flash/data/splits.py [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+        else:
+            indices = list(np.unique(indices))
+
+        if np.max(indices) >= len(dataset) or np.min(indices) < 0:
+            raise MisconfigurationException(f"`indices` should be within [0, {len(dataset) -1}].")
+
+        self.dataset = dataset
+        self.indices = indices
+
+    def __getitem__(self, index) -> Any:
```
Comment (ethanwharris) [inline flash/data/splits.py:41]:
```suggestion
    def __getitem__(self, index: int) -> Any:
```

File: flash/vision/classification/data.py [hunk 2]
```diff
@@ -49,18 +48,18 @@
...
-    image_size = (196, 196)
+    to_tensor = T.ToTensor()
 
     def __init__(
         self,
         train_transform: Optional[Union[Dict[str, Callable]]] = None,
         val_transform: Optional[Union[Dict[str, Callable]]] = None,
         test_transform: Optional[Union[Dict[str, Callable]]] = None,
         predict_transform: Optional[Union[Dict[str, Callable]]] = None,
+        image_size: Tuple[int, int] = (196, 196)
```
Comment (ethanwharris) [inline flash/vision/classification/data.py:59]:
```suggestion
        image_size: Tuple[int, int] = (196, 196),
```

File: flash/vision/classification/data.py [hunk 3]
```diff
@@ -96,8 +95,7 @@ def _get_predicting_files(samples: Union[Sequence, str]) -> List[str]:
 
         return files
 
-    def default_train_transforms(self):
-        image_size = self.image_size
+    def default_train_transforms(self, image_size):
```
Comment (ethanwharris) [inline flash/vision/classification/data.py:98]:
```suggestion
    def default_train_transforms(self, image_size: Tuple[int, int]):
```

File: flash/vision/classification/data.py [hunk 4]
```diff
@@ -117,8 +115,7 @@ def default_train_transforms(self):
                 "post_tensor_transform": T.Normalize([0.485, 0.456, 0.406], [0.229, 0.224, 0.225]),
             }
 
-    def default_val_transforms(self):
-        image_size = self.image_size
+    def default_val_transforms(self, image_size):
```
Comment (ethanwharris) [inline flash/vision/classification/data.py:118]:
```suggestion
    def default_val_transforms(self, image_size: Tuple[int, int]):
```

**Review by ethanwharris — APPROVED** [review]
Awesome, you even fixed the image_size pickle problem :smiley: Small typing misses

File: flash/data/splits.py (hunk 1 above)
Comment (tchaton) [reply flash/data/splits.py:41]: Done !

File: flash/vision/classification/data.py (hunk 2 above)
Comment (tchaton) [reply flash/vision/classification/data.py:59]: Done !

File: flash/vision/classification/data.py (hunk 3 above)
Comment (tchaton) [reply flash/vision/classification/data.py:98]: Done !

**Pull request merged**